src/modes/ModeEstimation.jl

```diff
@@ -0,0 +1,318 @@
+    vns, dist = get_vns_and_dist(right, left, vn)
+    r = get_and_set_val!(vi, vns, dist, sampler)
+    lp = sum(logpdf(dist, r))
+    return r, lp
+end
+
+# dot observe
+function DynamicPPL.dot_tilde(ctx::OptimizationContext{<:PriorContext}, sampler, right, left, vn, _, vi)
+    return 0
+end
+
+function DynamicPPL.dot_tilde(ctx::OptimizationContext{<:PriorContext}, sampler, right, left, vi)
+    return 0
+end
+
+function DynamicPPL.dot_tilde(ctx::OptimizationContext, sampler, right, left, vn, _, vi)
+    vns, dist = get_vns_and_dist(right, left, vn)
+    r = get_and_set_val!(vi, vns, dist, sampler)
+    lp = sum(logpdf(dist, r))
+    return lp
+end
+
+function DynamicPPL.dot_tilde(ctx::OptimizationContext, sampler, dists, value, vi)
+    return sum(Distributions.logpdf.(dists, value))
+end
+
+"""
+    OptimLogDensity{M<:Model,C<:Context,V<:VarInfo}
+
+A struct that stores the log density function of a `DynamicPPL` model.
+"""
+struct OptimLogDensity{M<:Model,C<:AbstractContext,V<:VarInfo}
+    "A `DynamicPPL.Model` constructed either with the `@model` macro or manually."
+    model::M
+    "A `DynamicPPL.AbstractContext` used to evaluate the model. `LikelihoodContext` or `DefaultContext` are typical for MAP/MLE."
+    context::C
+    "A `DynamicPPL.VarInfo` struct that will be used to update model parameters."
+    vi::V
+    "Whether the to evaluate the model with transformed variables."
+    linked::Bool
+end
+
+"""
+    OptimLogDensity(model::Model, context::AbstractContext)
+
+Create a callable `OptimLogDensity` struct that evaluates a model using the given `context`.
+"""
+function OptimLogDensity(model::Model, context::AbstractContext, linked::Bool)
+	init = VarInfo(model, context)
+	linked && DynamicPPL.link!(init, DynamicPPL.SampleFromPrior())
+	return OptimLogDensity(model, context, init, linked)
+end
+
+"""
+    (f::OptimLogDensity)(z; unlinked::Bool = false)
+
+Evaluate the log joint (with `DefaultContext`) or log likelihood (with `LikelihoodContext`)
+at the array `z`. If `unlinked=true`, no change of variables will occur.
+"""
+function (f::OptimLogDensity)(z; linked::Bool = true)
+    spl = DynamicPPL.SampleFromPrior()
+
+    varinfo = DynamicPPL.VarInfo(f.vi, spl, z)
+    f.model(varinfo, spl, f.context)
+    return -DynamicPPL.getlogp(varinfo)
+end
+
+"""
+    unlink(f::OptimLogDensity)
+
+Generate an unlinked (with no variable transformions) version of an existing `OptimLogDensity`.
+"""
+function Bijectors.invlink(f::OptimLogDensity)
+    init = VarInfo(f.model, f.context)
+    return OptimLogDensity(f.model, f.context, init, false)
+end
+
+"""
+    link(f::OptimLogDensity)
+
+Generate an linked (with variable transformions) version of an existing `OptimLogDensity`.
+"""
+function Bijectors.link(f::OptimLogDensity)
+    init = VarInfo(f.model, f.context)
+    DynamicPPL.link!(init, DynamicPPL.SampleFromPrior())
+    return OptimLogDensity(f.model, f.context, init, false)
+end
```

Comment:
I'm curious, why are `invlink` and `link` needed? BTW the docstring of `invlink` is incorrect.

Reply:
I don't think it's needed anymore with `OptimizationContext`, I'll remove it.